src/protocol/aggregation/mod.rs

```diff
@@ -52,139 +58,229 @@ where
-    let bits = contribution_value_bits_gf2[0].len();
-
-    // mod-convert for later validation
-    let convert_ctx = ctx
-        .narrow(&Step::ConvertValueBits)
-        .set_total_records(record_count);
-    let converted_contribution_values = convert_bits(
-        convert_ctx,
-        stream_iter(contribution_value_bits_gf2),
-        0..u32::try_from(bits).unwrap(),
-    );
+    debug_assert!(contribution_values.len() == breakdown_keys.len());
+    let num_records = contribution_values.len();
+    // for now, we assume that the bucket count is 2^BK::BITS
+    let num_buckets = 1 << breakdown_keys[0].len();
 
-    let aggregate = converted_contribution_values
-        .try_fold(S::ZERO, |mut acc, row| async move {
-            acc += &row.to_additive_sharing_in_large_field();
+    let equality_check_ctx = ctx
+        .narrow(&Step::ComputeEqualityChecks)
+        .set_total_records(num_records);
+    let check_times_value_ctx = ctx
+        .narrow(&Step::CheckTimesValue)
+        .set_total_records(num_records);
+
+    // Generate N streams for each bucket specified by the breakdown key (N = |breakdown_keys|).
+    // A stream is pipeline of contribution values multiplied by the "check bit". A check bit is
+    // a bit that is a share of 1 if the breakdown key matches the bucket, and 0 otherwise.
+    let streams = seq_join(
+        ctx.active_work(),
+        stream_iter(breakdown_keys)
+            .zip(stream_iter(contribution_values))
+            .enumerate()
+            .map(|(i, (bk, v))| {
+                let eq_ctx = &equality_check_ctx;
+                let mul_ctx = &check_times_value_ctx;
+                async move {
+                    let equality_checks = check_everything(eq_ctx.clone(), i, &bk).await?;
+                    let value = &v;
+                    eq_ctx
+                        .try_join(
+                            equality_checks
+                                .into_iter()
+                                .take(num_buckets)
+                                .enumerate()
+                                .map(|(check_idx, check)| {
+                                    let step = BitOpStep::from(check_idx);
+                                    let c = mul_ctx.narrow(&step);
+                                    let record_id = RecordId::from(i);
+                                    async move {
+                                        check
+                                            .multiply(
+                                                &value.to_additive_sharing_in_large_field(),
+                                                c,
+                                                record_id,
+                                            )
+                                            .await
+                                    }
+                                }),
+                        )
+                        .await
+                }
```

Comment:
I find this really difficult to read / digest / understand.

Is there anything we can do to improve the code readability here? Perhaps something that leads to less indentation? Maybe factor something out into another function?

Reply:
Extracted the multiplication part.